services/graph/pkg/service/v0/base.go

```diff
@@ -0,0 +1,692 @@
+	} else {
+		actions := unifiedrole.CS3ResourcePermissionsToLibregraphActions(*share.GetPermissions().GetPermissions())
+		perm.SetLibreGraphPermissionsActions(actions)
+		perm.SetRoles(nil)
+	}
+	perm.SetGrantedToV2(grantedTo)
+	return &perm, nil
+}
+
+func (g BaseGraphService) cs3PublicSharesToDriveItems(ctx context.Context, shares []*link.PublicShare, driveItems driveItemsByResourceID) (driveItemsByResourceID, error) {
+	for _, s := range shares {
+		g.logger.Debug().Interface("CS3 PublicShare", s).Msg("Got Share")
+		resIDStr := storagespace.FormatResourceID(*s.ResourceId)
+		item, ok := driveItems[resIDStr]
+		if !ok {
+			itemptr, err := g.getDriveItem(ctx, storageprovider.Reference{ResourceId: s.ResourceId})
+			if err != nil {
+				g.logger.Debug().Err(err).Interface("Share", s.ResourceId).Msg("could not stat share, skipping")
+				continue
+			}
+			item = *itemptr
+		}
+		perm, err := g.libreGraphPermissionFromCS3PublicShare(s)
+		if err != nil {
+			g.logger.Error().Err(err).Interface("Link", s.ResourceId).Msg("could not convert link to libregraph")
+			return driveItems, err
+		}
+
+		item.Permissions = append(item.Permissions, *perm)
+		driveItems[resIDStr] = item
+	}
+
+	return driveItems, nil
+}
+
+func (g BaseGraphService) getLinkPermissionResourceID(ctx context.Context, permissionID string) (*storageprovider.ResourceId, error) {
+	share, err := g.getCS3PublicShareByID(ctx, permissionID)
+	if err != nil {
+		return nil, err
+	}
+	return share.GetResourceId(), nil
+}
+
+func (g BaseGraphService) getCS3PublicShareByID(ctx context.Context, permissionID string) (*link.PublicShare, error) {
+	gatewayClient, err := g.gatewaySelector.Next()
+	if err != nil {
+		g.logger.Debug().Err(err).Msg("selecting gatewaySelector failed")
+		return nil, err
+	}
+
+	getPublicShareResp, err := gatewayClient.GetPublicShare(ctx,
+		&link.GetPublicShareRequest{
+			Ref: &link.PublicShareReference{
+				Spec: &link.PublicShareReference_Id{
+					Id: &link.PublicShareId{
+						OpaqueId: permissionID,
+					},
+				},
+			},
+		},
+	)
+	if errCode := errorcode.FromCS3Status(getPublicShareResp.GetStatus(), err); errCode != nil {
+		return nil, *errCode
+	}
+	return getPublicShareResp.GetShare(), nil
+}
+
+func (g BaseGraphService) removePublicShare(ctx context.Context, permissionID string) error {
+	gatewayClient, err := g.gatewaySelector.Next()
+	if err != nil {
+		g.logger.Debug().Err(err).Msg("selecting gatewaySelector failed")
+		return err
+	}
+
+	removePublicShareResp, err := gatewayClient.RemovePublicShare(ctx,
+		&link.RemovePublicShareRequest{
+			Ref: &link.PublicShareReference{
+				Spec: &link.PublicShareReference_Id{
+					Id: &link.PublicShareId{
+						OpaqueId: permissionID,
+					},
+				},
+			},
+		})
+	if errcode := errorcode.FromCS3Status(removePublicShareResp.GetStatus(), err); errcode != nil {
+		return *errcode
+	}
+	// We need to return an untyped nil here otherwise the error==nil check won't work
```

Comment:
Which `error==nil` check? And why?

Reply:
If we'd just do:

```
return errorcode.FromCS3Status(removePublicShareResp.GetStatus(), err)
```

here, the caller (of  getCS3PublicShareByID() in this case) could not simply check from `err==nil`. Because the returned value would be a "typed nil". Basically https://go.dev/doc/faq#nil_error

I think the root cause is in the way the errorcode package is implemented. We need to clean that up  at some point as well.